posts/2021/10/array-libraries-interoperability.md

```diff
@@ -0,0 +1,600 @@
+own methods which conventionally ingest arrays from these array provider
+libraries as their arguments. We'll refer to these set of libraries as **"array
+consumer"** libraries.
+
+
+## What's the catch?
+
+The pain now is that an extremely stable and popular consumer library
+like SciPy is cut off from the sweet features which the new array/tensor
+frameworks have to offer. A user might prefer SciPy's domain specific
+features and stable API for their scientific stack, but because their
+underlying project is built on top of say PyTorch, they are left alone
+at sea.
+
+Say we wanted to use a provider library other than NumPy, but with SciPy
+(or a similar library). After some thought, one may end up choosing
+either of the following options:
+
+1. Somehow convert their array/tensors into NumPy arrays,
+	before feeding them in the SciPy function, which is where you lose
+	the *"extra sweetness"*.
+
+2. Find some other consumer library (if it exists) mimicking SciPy's API, but
+	built on top of the "array provider" library that was used in the initial
+	project (say PyTorch). This way the user enjoys the extra
+	features, but this comes at the cost of learning and getting used to
+	a new library API which might mimic SciPy's functionality
+	but not behave exactly the same.
+
+
+Also, the APIs (function signatures) for these "array providers" can
+be different, which is a problem in itself. Imagine a
+scientist who has experience using NumPy for about 10 years. When they
+move into the PyTorch ecosystem, they are still doing a lot of similar
+computations, but now will need to learn a new API.
+
+
+<p align="center">
+	<img
+     alt="Comic depicting the happy relationship between NumPy and SciPy, and
+     how envious other array/tensor libraries are of it."
+     src="/images/2021/10/array_wonderland.JPG"
+	 style="object-fit:cover;
+            width: 80%;"/>
+</p>
+
+<i align="center"><b>ELI5 Caption:</b> This is a page from the array libraries interoperability special edition chapter of comic array wonderland. SciPy only wants to play with NumPy and refuses to play with other libraries like PyTorch, CuPy, Tensorflow and JAX. All of them are sad, and then they decide to make their own new friends.
+</i>
+
+<i align="center"><b>Serious Caption:</b> Other than NumPy, array/tensor libraries like PyTorch, CuPy, Tensorflow and JAX aren't compatible with Consumer Libraries like SciPy. This led to developers creating an ecosystem of libraries around each array/tensor framework which are conceptually the same but differ in the unique array/tensor frameworks they are built for.
+</i>
+
+---
+
+Life is hard, ain't it! But as Rocky Balboa said:
+> "NumPy, CuPy, PyTorch or SciPy is not gonna hit as hard as all of them
+>  when used together. But it ain't about finding a way to use them
+>  individually; it's about making them work together.
+>  That's how interoperable science is done." ~ ([actual quote :P](https://youtu.be/8xFEqdkO5UI?t=13))
```

Comment:
```suggestion
>  That's how data science is done." ~ ([actual quote :P](https://youtu.be/8xFEqdkO5UI?t=13))
```

This quote really doesn't mean anything, it is just my imagination if Rocky were to talk about interoperability. But for consistency I'll remove this "interoperable science" phrase.

Reply:
sounds good